cpp/src/Ice/ConnectionI.cpp

```diff
@@ -394,9 +403,10 @@ Ice::ConnectionI::start(const StartCallbackPtr& callback)
 
         if(!initialize() || !validate())
         {
-            if(callback)
+            if(connectionStartCompleted && connectionStartFailed)
```

Comment:
It would be nicer if connectionStartCompleted and connectionStartFailed were always callable (not null), but that's not how the old code is setup.